### PR TITLE
feat(cleaner): write unit test for cleaner 

### PR DESCRIPTION
**Pull request by smothiki**
Little tricky as my first run got hung up at watching events at the channel
Working more on this.

### CODE REVIEW

File: pkg/cleaner/cleaner.go [hunk 1]
```diff
@@ -3,20 +3,19 @@ package cleaner
...
 )
 
 const (
 	dotGitSuffix = ".git"
 )
 
-// Run starts the deleted app cleaner. Every pollSleepDuration, it compares the result of nsLister.List with the directories in the top level of gitHome on the local file system. On any error, it uses log messages to output a human readable description of what happened.
-func Run(gitHome string, nsLister k8s.NamespaceWatcher) error {
-
+// Run starts the deleted app cleaner. Every pollSleepD uration, it compares the result of nsLister.List with the directories in the top level of gitHome on the local file system. On any error, it uses log messages to output a human readable description of what happened.
```
Comment (arschles) [inline pkg/cleaner/cleaner.go]: @smothiki these docs are out of date now. Can you fix them to reflect the fact that it listens to the event stream. Maybe something like this:

```
Run starts the deleted app cleaner. This function listens to the Kubernetes event stream for events that indicate namespaces that were `DELETED`. If the namespace name matches a folder on the local filesystem, this func deletes that folder. Note that this function blocks until the watcher returned by `nsLister.Watch` is closed, so you should launch it in a goroutine.
```

File: pkg/cleaner/cleaner.go [hunk 2]
```diff
@@ -26,7 +25,7 @@ func Run(gitHome string, nsLister k8s.NamespaceWatcher) error {
 		if event.Type == "DELETED" {
 			namespace := event.Object.(*api.Namespace)
```
Comment (arschles) [inline pkg/cleaner/cleaner.go]: @smothiki you should check this type assertion, and if it wasn't an `*api.Namespace`, then don't continue.

File: pkg/cleaner/cleaner_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	fs := sys.NewFakeFS()
+	dirhome := "/home/git"
+	fs.Files["/home/git/dir1.git"] = []byte{}
+	go Run(dirhome, ns, fs)
+	time.Sleep(5 * time.Second)
+	_, ok := fs.Files["/home/git/dir1.git"]
+	if ok {
+		t.Fatal("direcotry not deleted unit test failed")
+	}
+}
```
Comment (arschles) [inline pkg/cleaner/cleaner_test.go]: @smothiki can you add a few other tests for failure cases. For example:

1. The watcher sends other objects, like pods or replication controllers. In those cases, the cleaner should do nothing
2. The watcher sends namespaces that don't exist in the fake filesystem
3. The watcher doesn't send any `DELETE` events

File: pkg/cleaner/cleaner_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type namespace struct {
+}
+
+// Watch returns a watch.Interface that watches the requested namespaces.
+
```
Comment (arschles) [inline pkg/cleaner/cleaner_test.go]: nitpick: can you remove this line break

File: pkg/cleaner/cleaner_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	go func() {
+		nst.Add(&api.Namespace{ObjectMeta: api.ObjectMeta{Name: "dir1"}})
+		nst.Modify(&api.Namespace{ObjectMeta: api.ObjectMeta{Name: "dir1"}})
+		nst.Delete(&api.Namespace{ObjectMeta: api.ObjectMeta{Name: "dir1"}})
+		nst.Stop()
+	}()
+	return nst, nil
+}
+
+func (r *namespace) IsAnAPIObject() {}
```
Comment (arschles) [inline pkg/cleaner/cleaner_test.go]: doesn't this need to return `bool`?

**Comment by smothiki** [comment]
@arschles thanks for the comments . I'm updating the PR accordingly

File: pkg/cleaner/cleaner_test.go (hunk 5 above)
Comment (smothiki) [inline pkg/cleaner/cleaner_test.go]: IsANAPIObject doesn;t return anything. It just has a pointer receiver

**Comment by arschles** [comment]
@smothiki thanks - let me know when you'd like another review

**Comment by smothiki** [comment]
@arschles  let me know if we can delete testdata dir

**Comment by arschles** [comment]
@smothiki I reviewed again, and only [this comment](https://github.com/deis/builder/pull/242#discussion_r55739907) remains. Also, is the testdata directory used anywhere else?

**Pull request merged**